### PR TITLE
executors: Cleanup nightly AMI images after 60 days

### PR DESCRIPTION
**Pull request by sanderginn**
- Closes https://github.com/sourcegraph/sourcegraph/issues/46786

This PR adds a post-processor to the Packer builds of the executor and docker mirror images that looks for images older than (currently configured) 60 days and deletes them.

Steps taken to get this to work:
- Added a tag `Amazon_AMI_Management_Identifier` with a value equal to the base name of the nightly builds (`sourcegraph-executors-nightly` and `sourcegraph-executors-docker-mirror-nightly`) to the respective images:
  - Get all AMIs for the executor image:
   `IMAGE_IDS=$(aws ec2 --region=us-west-2 describe-images --filter "Name=name,Values=sourcegraph-executors-nightly*" --query 'Images[*].ImageId' --output json | jq -r 'join(" ")')`
  - For those AMIs, run
   `aws ec2 create-tags --region us-west-2 --resources $IMAGE_IDS --tags Key=Amazon_AMI_Management_Identifier,Value=sourcegraph-executors-nightly`
  - Repeat for the K/V pair `sourcegraph-executors-docker-mirror-nightly`
- Added the following permissions to the IAM policy for the AWS account used by Buildkite:
   ```json
   "autoscaling:DescribeLaunchConfigurations",
   "ec2:DescribeLaunchTemplates",
   "ec2:DescribeLaunchTemplateVersions",
   ```

### Todo 
- [ ] Document this behavior somewhere
- [ ] Remove `dry_run` flag from both Packer configs (after approval, before merging)

## Test plan
Ran with `dry_run = true` enabled on the post-processor. This printed a big list, starting with the following IDs (the `null.basic-example` was a no-op build source for speedier local testing):
```shell
null.basic-example (amazon-ami-management): [DryRun] Deleting image: ami-0b330916df93dd142
null.basic-example (amazon-ami-management): [DryRun] Deleting image: ami-00347265347e36825
null.basic-example (amazon-ami-management): [DryRun] Deleting image: ami-0b1e0f9a859054435
```
As shown in the image, the first ID is for an AMI with a creation date <60 days ago at the time of running (April 3rd), whereas from the second AMI and onward the creation date is >60 days ago, and are selected for deletion.
Screenshot 2023-04-03 at 21 27 28

Furthermore, the post-processor will _not_ delete any AMIs that are in use by any existing configurations, even earlier versions. For example, this is a launch template that uses an AMI in version 9 of the template, while it is currently on version 13:
```shell
null.basic-example (amazon-ami-management): [WARN] ami-05280d193532cd37e is used in launch template: sourcegraph_executor-template-20221110203635535000000005 (9). Skipped...
```

### CODE REVIEW

File: enterprise/cmd/executor/vm-image/executor.pkr.hcl [hunk 1]
```diff
@@ -137,4 +150,14 @@ build {
             }
         }
     }
+
+    post-processor "amazon-ami-management" {
+        only       = ["amazon-ebs.aws"]
+        access_key = var.aws_access_key
+        secret_key = var.aws_secret_key
+        regions    = ["us-west-2"]
+        identifier = var.image_family
```
[the remark this answers is not in the document]
Comment (sanderginn) [reply enterprise/cmd/executor/vm-image/executor.pkr.hcl:166]: Good point, I'm doing a test run that sets the identifier only if `EXECUTOR_IS_TAGGED_RELEASE` is false.

[the remark this answers is not in the document]
Comment (sanderginn) [reply enterprise/cmd/executor/vm-image/executor.pkr.hcl:166]: I forced `EXECUTOR_IS_TAGGED_RELEASE` to true for this build, no identifier tag included 😄 

```shell
==> amazon-ebs.aws: Creating AMI tags
    amazon-ebs.aws: Adding tag: "Base_AMI_Name": "ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20230328"
    amazon-ebs.aws: Adding tag: "Extra": "<no value>"
    amazon-ebs.aws: Adding tag: "Name": "sourcegraph-executors-nightly-211722"
    amazon-ebs.aws: Adding tag: "OS_Version": "Ubuntu"
    amazon-ebs.aws: Adding tag: "Release": "Latest"
```

[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/cmd/executor/vm-image/executor.pkr.hcl:166]: Nice!

**Review by eseliger — APPROVED** [review]
wonder if the same could be done for GCP as well

**Comment by sanderginn** [comment]
@eseliger FYI https://github.com/sourcegraph/sourcegraph/issues/50334

**Comment by eseliger** [comment]
Nice thanks!

**Pull request merged**